File: online_store/domain/models/products/Book.py
from online_store.domain.models.products.ProductInterface import ProductInterface

class Book(ProductInterface):
    def __init__(self, title: str, author: str, price: float) -> None:
        self.title = title
        self.author = author
        self.price = price

    def getPrice(self) -> float:
        return self.price
    
    def modifyPrice(self, new_price:float):
        self.price = new_price

    def getTitle(self) -> str:
        return self.title
    
    def clone(self):
        return Book(self.title, self.author, self.price)
    
    def __str__(self):
        return f"{self.title} by {self.author} - Price: ${self.price:.2f}"

File: tmpsjava/online_shop/src/domain/models/products/Book.java
package domain.models.products;

public class Book implements ProductInterface{
    final private String title;
    private float price;

    private Promotion promotion;

    public Book(String title, float price, Promotion promotion){
        this.title = title;
        this.price = price;
        this.promotion = promotion;
    }

    @Override
    public float getPrice() {
        return this.promotion.applyDiscount(this.price);
    }

    @Override
    public void modifyPrice(float new_price) {
        this.price = new_price;
    }

    @Override
    public String getTitle() {
        return this.title;
    }

    @Override
    public ProductInterface copy() {
        return (ProductInterface) new Book(this.title, this.price, this.promotion);
    }

    @Override
    public Promotion getPromotion() {
        return promotion;
    }
}
